### PR TITLE
refactor/coinzoom api throttler

### PR DESCRIPTION
**Pull request by aarmoa**
**Before submitting this PR, please make sure**:

- [x] Your code builds clean without any errors or warnings
- [x] You are using approved title ("feat/", "fix/", "docs/", "refactor/")

**A description of the changes proposed in the pull request**:
Changes in Coinzoom connector to stop using the old throttler and replace it with an instance of the new AsyncThrottler.

**Tests performed by the developer**:
Added unit tests for all the functions that use the throttler to send http requests or websocket requests.

**Tips for QA testing**:
For this PR we require the basic connector validations.

### CODE REVIEW

**Comment by dennisocana** [comment]
We still need to merge the API Throttler integration for KuCoin to easily review this pull request.

**Comment by aarmoa** [comment]
The PR can now be reviewed (both from development side and from QA side)

**Comment by PtrckM** [comment]
- keep getting this error frequently after starting the bot,  `rate_limits_share_pct 100` on `order_refresh_time 20` and disabled `order_refresh_tolerance_pct`. No order stuck and does not seems to affect creating and cancelling orders. After sometime the error message disappears.

orderbook_error

```
2021-09-11 04:24:14,729 - 63959 - hummingbot.connector.exchange.coinzoom.coinzoom_order_book_tracker - NETWORK - Unexpected error routing order book messages.
Traceback (most recent call last):
  File "/home/ubuntu/coinzoom_api_throttler/hummingbot/core/data_type/order_book_tracker.py", line 301, in _emit_trade_event_loop
    trade_message.content["trade_type"] == float(TradeType.SELL.value) else TradeType.BUY
KeyError: 'trade_type'
```
- getting random reconnection/disconnection when api rate limit warning shows causing timedout error, no order getting stuck on exchange as well.
timedout

```
2021-09-11 04:47:56,004 - 64113 - hummingbot.core.utils.async_utils - ERROR - Unhandled error in background task:
Traceback (most recent call last):
  File "hummingbot/connector/in_flight_order_base.pyx", line 84, in hummingbot.connector.in_flight_order_base.InFlightOrderBase.get_exchange_order_id
    await self.exchange_order_id_update_event.wait()
  File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/asyncio/locks.py", line 309, in wait
    await fut
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/ubuntu/coinzoom_api_throttler/hummingbot/core/utils/async_utils.py", line 9, in safe_wrapper
    return await c
  File "/home/ubuntu/coinzoom_api_throttler/hummingbot/connector/exchange/coinzoom/coinzoom_exchange.py", line 532, in _execute_cancel
    await tracked_order.get_exchange_order_id()
  File "hummingbot/connector/in_flight_order_base.pyx", line 83, in get_exchange_order_id
    async with timeout(10):
  File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/site-packages/async_timeout/__init__.py", line 55, in __aexit__
    self._do_exit(exc_type)
  File "/home/ubuntu/miniconda3/envs/hummingbot/lib/python3.8/site-packages/async_timeout/__init__.py", line 92, in _do_exit
    raise asyncio.TimeoutError
asyncio.exceptions.TimeoutError
```

- config and logs:
[cz.zip](https://github.com/CoinAlpha/hummingbot/files/7147220/cz.zip)

File: test/hummingbot/connector/exchange/coinzoom/test_coinzoom_api_order_book_data_source.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+            websocket_mock=ws_connect_mock.return_value,
+            message=json.dumps(message))
+        trade_message = self.async_run_with_timeout(received_messages.get())
+
+        self.assertEqual(OrderBookMessageType.TRADE, trade_message.type)
+        self.assertEqual(int(dateparse("2020-01-16T21:02:23Z").timestamp() * 1e3), trade_message.timestamp)
+        self.assertEqual(trade_message.timestamp, trade_message.trade_id)
+        self.assertEqual(self.trading_pair, trade_message.trading_pair)
+
+    @patch("time.time")
```
Comment (petioptrv) [inline test/hummingbot/connector/exchange/coinzoom/test_coinzoom_api_order_book_data_source.py:111]: Shouldn't we, in this case, apply the same treatment as with the `asyncio.sleep` module; i.e. mock a module-specific function instead?

File: test/hummingbot/connector/exchange/coinzoom/test_coinzoom_api_order_book_data_source.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    def tearDown(self) -> None:
+        self.listening_task and self.listening_task.cancel()
+        super().tearDown()
+
+    def async_run_with_timeout(self, coroutine: Awaitable, timeout: int = 1):
+        ret = asyncio.get_event_loop().run_until_complete(asyncio.wait_for(coroutine, timeout))
+        return ret
+
+    @staticmethod
+    def _register_sent_request(requests_list, url, **kwargs):
```
Comment (petioptrv) [inline test/hummingbot/connector/exchange/coinzoom/test_coinzoom_api_order_book_data_source.py:48]: Unused method

**Review by petioptrv — CHANGES_REQUESTED** [review]
Left a couple comments in the code. There is also an edge-case scenarios that I encountered in the previous adaptations that I think is not yet handled here: the possibility of an order-creation being rate-limited, then calling `_execute_cancel` which can timeout on line 532 (`await tracked_order.get_exchange_order_id()`). The previous adaptations just skip the cancellation request, which I think here would amount to setting `order_was_cancelled = False`. The strategy will request the cancellation again on the next iteration and if the order has been submitted, things will go as planned.

**Comment by aarmoa** [comment]
@PatrickLiPM thank you for the details. The issue should be solved with the last commit.

**Review by petioptrv — APPROVED** [review]
I like the idea to make the order non-local in `update_exchange_order_id`. I'll adopt that for Kraken as well

**Review by rxlxrxsx — APPROVED** [review]
Confirmed fix for issues listed above and no other issues found during testing by QA. Approved.

**Pull request merged**